beacon-chain/p2p/peers/score_block_providers.go

```diff
@@ -0,0 +1,278 @@
+	// StartScore defines initial score from which peer starts. Set to positive to give peers an
+	// opportunity to be selected for block fetching (allows new peers to start participating,
+	// when there are already scored peers).
+	StartScore float64
+	// ReturnedBlocksWeight defines weight of a returned/requested ratio in overall score.
+	ReturnedBlocksWeight float64
+	// SlowReturnedBlocksPenalty defines a penalty applied to score, if blocks were requested,
+	// but none have been returned yet (to distinguish between non-responsive peers and peers that
+	// haven't been requested any blocks yet).
+	SlowReturnedBlocksPenalty float64
+	// ProcessedBlocksWeight defines weight of a processed/requested ratio in overall score.
+	ProcessedBlocksWeight float64
+	// SlowProcessedBlocksPenalty defines a penalty applied to score, if blocks have been
+	// requested and returned, but none have been processed yet. Allows distinguishing between
+	// peers that haven't yet returned anything and peers that returned garbage.
+	SlowProcessedBlocksPenalty float64
+	// DecayInterval defines how often requested/returned/processed stats should be decayed.
+	DecayInterval time.Duration
+	// Decay specifies the factor (must be < 1.0) by which block provider's stats is decayed.
+	Decay float64
+}
+
+// newBlockProviderScorer creates block provider scoring service.
+func newBlockProviderScorer(
+	ctx context.Context, store *peerDataStore, config *BlockProviderScorerConfig) *BlockProviderScorer {
+	if config == nil {
+		config = &BlockProviderScorerConfig{}
+	}
+	scorer := &BlockProviderScorer{
+		ctx:    ctx,
+		config: config,
+		store:  store,
+	}
+	if scorer.config.StartScore == 0.0 {
+		scorer.config.StartScore = DefaultBlockProviderStartScore
+	}
+	if scorer.config.ReturnedBlocksWeight == 0.0 {
+		scorer.config.ReturnedBlocksWeight = DefaultBlockProviderReturnedBlocksWeight
+	}
+	if scorer.config.SlowReturnedBlocksPenalty == 0.0 {
+		scorer.config.SlowReturnedBlocksPenalty = DefaultSlowReturnedBlocksPenalty
+	}
+	if scorer.config.ProcessedBlocksWeight == 0.0 {
+		scorer.config.ProcessedBlocksWeight = DefaultBlockProviderProcessedBlocksWeight
+	}
+	if scorer.config.SlowProcessedBlocksPenalty == 0.0 {
+		scorer.config.SlowProcessedBlocksPenalty = DefaultSlowProcessedBlocksPenalty
+	}
+	if scorer.config.DecayInterval == 0 {
+		scorer.config.DecayInterval = DefaultBlockProviderDecayInterval
+	}
+	if scorer.config.Decay == 0.0 {
+		scorer.config.Decay = DefaultBlockProviderDecay
+	}
+	return scorer
+}
+
+// Score calculates and returns total score based on returned and processed blocks.
+func (s *BlockProviderScorer) Score(pid peer.ID) float64 {
+	s.store.RLock()
+	defer s.store.RUnlock()
+	return s.score(pid)
+}
+
+// score is a lock-free version of Score.
+func (s *BlockProviderScorer) score(pid peer.ID) float64 {
+	score := s.Params().StartScore
+	peerData, ok := s.store.peers[pid]
+	if ok && peerData.requestedBlocks > 0 {
+		// Score returned/requested ratio.
```

Comment:
First of all, please note, that `BlockProviderScore` is not designed for disconnecting and banning peers, it is more of a tool used to sort and select the most viable peers (based on the latest performance). So, in case of a peer unfairly penalized, the stats constantly decays (with a current decay rate of `-5%` every minute, it takes half an hour to decrease stats by 80%).

That being said, skipped slots are accounted for the following way:
- In `init-sync` PR (#6709) I do not use block provider scorer when searching for non-skipped slot -- because algorithm is deliberately probing wide range of slot numbers, and it would be not ok to update score. Still non-skipped slot searcher relies on score to try and use most viable peer.
- To combat temporary unfairness, non-zero start score exits: so peer starts from `0.1` (configurable) and can tolerate the damage taken when penalized due to inactivity (the total penalty is `-0.1`, so peer is still viable, just not performant).
- The `init-sync` will use 75% of ordered peers (ordered by score). Then when selecting out of those 75% peers, capacity is taken into account (currently it accounts for 30% of score - weight is configurable). So, if higher scored peers exhausted their capacity (relatively), peer will have a chance to fetch the blocks.

So, overall, it doesn't matter when on `BeaconBlocksByRange` and peer responds with empty list, whether this happened because peer has temporary difficulties, is malicious, there are truly skipped slots -- as block provider it is considered less effective that peer able to produce blocks. Then, since not only produced but processed blocks affect the score, peer break ties by giving more blocks that score successfully. If we are within a region of big non-finality, then **all** peers will be hit by the algorithm, and since block provider scores are used to weighted round robin selection (and not for banning) + we have decaying -- that's not the problem.

Reply:
I guess my main issue with using `returned/requested` is which peer quality we are exactly scoring. While this may not lead to peer bans/disconnects, this metric is not very helpful when trying to score the average peer. Of the blocks returned are those blocks of the right quality ? 

A peer could return us 64 blocks, but the list could be non-linear, or generally not very good. I do also know that you are using processed/returned as a ratio too to assess peer quality. While that is fine usually, how well a batch is processed is dependent on the previous peer in block queue. Ex: 
```
Peer B -> Returns 64 blocks (linear list)

Peer A -> Returns 63 blocks ( Also linear list, but last block is missing)
```
You would end up with peer B being penalised more harshly since their whole batch is marked as bad.  I understand this is more of a statistical guarantee that we will be able to be mark down outliers while keeping our sync humming along. However how well does this play with low peer counts ? Ex: I only have 3 peers, 1 bad peer could potentially mark down the score of the other 2 peers long enough. As the decay takes about 30 minutes, you would have that lousy peer just giving us bad blocks.